From 3f420a49dd4d0c413b635dd621aa34eebce2d3d2 Mon Sep 17 00:00:00 2001
From: Karel Zak <kzak@redhat.com>
Date: Mon, 5 Aug 2013 13:58:01 +0200
Subject: [PATCH] libmount: canonicalize for conversion from loopdev backing
 file

  # mount foo.img /mnt
  # umount foo.img
  umount: foo.img: not mounted

The loopdev code (and sysfs backing_file) uses absolute paths, but
libmount does not canonicalize the path before lookup for the backing file.

References: https://bugzilla.redhat.com/show_bug.cgi?id=950497
Signed-off-by: Karel Zak <kzak@redhat.com>
---
 libmount/src/context_umount.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/libmount/src/context_umount.c b/libmount/src/context_umount.c
index 2b791c4..b02902c 100644
--- a/libmount/src/context_umount.c
+++ b/libmount/src/context_umount.c
@@ -161,7 +161,12 @@ try_loopdev:
		struct stat st;

		if (stat(tgt, &st) == 0 && S_ISREG(st.st_mode)) {
-			int count = loopdev_count_by_backing_file(tgt, &loopdev);
+			int count;
+
+			cn_tgt = mnt_resolve_path(tgt, cache);
+			count = loopdev_count_by_backing_file(cn_tgt, &loopdev);
+			if (!cache)
+				free(cn_tgt);
			if (count == 1) {
				DBG(CXT, mnt_debug_h(cxt,
					"umount: %s --> %s (retry)", tgt, loopdev));
--
1.8.1.4
